Add `pixi self-update`

hadim opened the pull request:
Supersede #582 

---

You can test it locally in your dev env with:

```bash
$ pixi run cargo run -- self-update --force --version 0.8.0
$ ./target/debug/pixi --version
pixi 0.8.0
```

---

See the current API:

```bash
$ pixi self-update
✔ Pixi will be updated from '0.1.0' to '0.12.0'
✔ Pixi archive downloaded.
✔ Pixi archive uncompressed.
✔ Pixi has been updated to version 0.12.0.
```

```bash
$ pixi self-update # wrong pixi location
✘ pixi is not installed in the default location:
- Default pixi location: /Users/hadim/.pixi/bin/pixi
- Pixi location detected: /Users/hadim/Code/libs/pixi/pixi/target/debug/pixi

It can happen when pixi has been installed via a dedicated package manager (such as Homebrew on macOS).
You can always use `pixi self-update --force` to force the update.
```

```bash
$ pixi self-update --version 0.8.0
✔ Pixi will be updated from '0.1.0' to '0.8.0'
✔ Pixi archive downloaded.
✔ Pixi archive uncompressed.
✔ Pixi has been updated to version 0.8.0.
```

```bash
$ pixi self-update
✔ pixi is already up-to-date (version 0.12.0)
```

```bash
$ pixi self-update --version 0.8.99
✘ The version you specified is not available: 0.8.99
```

wolfv commented on src/cli/self_update.rs:
```diff
@@ -0,0 +1,276 @@
+        .unwrap()
+        .join(".pixi")
+        .join("bin")
+        .join("pixi")
```
not sure if this needs to be `pixi.exe` on Windows. I can check

hadim replied:
Ah good point. I don't have a windows dev machine. I'll wait for your feedback.

hadim replied:
Hum the check is done with `starts_with` so I guess it should be fine:

```rust
    std::env::current_exe()
        .unwrap()
        .to_str()
        .unwrap()
        .starts_with(default_binary_path.to_str().unwrap())
```

wolfv replied:
yeah, it's pixi.exe. The `starts_with` on Path/ PathBuf compares the components so I don't think this would work :)

hadim replied:
i fixed it - should be fine now.

hadim commented on src/cli/self_update.rs:
```diff
@@ -0,0 +1,276 @@
+    // If args.force is false and pixi is not installed in the default location, stop here.
+    match (args.force, is_pixi_binary_default_location()) {
+        (false, false) => {
+            println!(
```
I suggest we go with that logic for now as it will work 99% of the time for the downstream users, and it's always possible to adapt the logic in the future if you get user's reports about edge cases.

Meanwhile, the `--force` flag will always allow people to update.

That being said, I am fine modifying it if you prefer.

hadim commented on src/cli/self_update.rs:
```diff
@@ -0,0 +1,276 @@
+            console::style(console::Emoji("✘ ", "")).red(),
+            std::env::consts::OS
+        );
+        std::process::exit(1);
```
I am using `std::process::exit(1);` in a couple of places in the code.

Is that ok with you? I am still struggling to find the best error handling logic. I tried a bit with `miette` but could not make it work.

Also should the exit code be `0`? or something else?

baszalmstra replied:
Yeah I think it would be better to just return a miette Error instead.

hadim replied:
I used `Err(miette::miette!(error_mesage))?` - it's indeed much better!

hadim commented on src/cli/self_update.rs:
```diff
@@ -0,0 +1,276 @@
+    // let current_version = "0.1.0"; // NOTE(hadim): for dev purposes
+
+    // Stop here if the target version is the same as the current version
+    if target_version == current_version {
```
The version comparison is basic, but I think it should be fine:

- it's simple (simpler is better xD)
- It allows downgrading the pixi version
- at this point in the code we already know the target version is valid and exists.

Let me know if you prefer something else.

wolfv commented:
I am happy with the logic. I'll let Bas and Ruben chime in with their thoughts!

baszalmstra commented on src/cli/self_update.rs:
```diff
@@ -0,0 +1,281 @@
+            std::process::exit(1);
+        }
+    }
+}
```
You can use https://doc.rust-lang.org/std/env/consts/constant.EXE_EXTENSION.html

hadim replied:
good catch - will use `std::env::consts::EXE_SUFFIX`

hadim commented:
@baszalmstra ready for another round!

I removed all the `unwrap` and replaced most of them by `except` (when the error is quite unlikely to happen).

I am also now using `miette` for "expected" errors.

I also replaced all the `println` to `eprintln`.

hadim commented:
> I also replaced all the println to eprintln.

I am actually unsure whether those should be printed to stdout or stderr looking at other cli commands.

baszalmstra reviewed (commented):
Some small nitpicks but looks good to me!

hadim commented on src/cli/self_update.rs:
```diff
@@ -0,0 +1,254 @@
+    // If args.force is false and pixi is not installed in the default location, stop here.
+    match (args.force, is_pixi_binary_default_location()) {
+        (false, false) => {
+            let error_mesage = format!(
```
Thanks.

hadim commented on src/cli/self_update.rs:
```diff
@@ -0,0 +1,254 @@
+                default_pixi_binary_path().to_str().expect("Could not convert the default pixi binary path to a string"),
+                env::current_exe().expect("Failed to retrieve the current pixi binary path").to_str().expect("Could not convert the current pixi binary path to a string")
+            );
+            Err(miette::miette!(error_mesage))?
```
Good to know.

hadim commented:
Thanks @baszalmstra - should be good now.

hadim commented:
> @hadim maybe you can follow up with a few docs. Great feature! :)

I already added the cli doc at `docs/cli.md` in that PR. Do you mean something else?

wolfv commented:
perfect, nope. sorry. didn't look closely enough 🤦 (getting late over here haha)
